File: integrations/langgraph/examples/03_multi_user_graph.py
"""One graph, many users — a hand-built StateGraph with templated namespaces.

Two things this shows that the prebuilt-agent example does not:

1. **The current, non-deprecated LangGraph API.** `create_react_agent` moved to
   `langchain.agents.create_agent` in LangGraph 1.0; a plain `StateGraph` never went anywhere, and
   `CortadelMemory`'s nodes are ordinary node callables, so they drop straight in.
2. **Multi-tenancy.** The namespace is a template, `("memories", "{user_id}")`, resolved per
   invocation from `config["configurable"]`. `namespace_user_id(-1)` then turns the resolved
   namespace back into the Cortadel user id, and the store keeps one client per user — which is
   required, because a Cortadel client is bound to a single user at construction.

Run it:

    docker compose up                      # from the cortadel repo root -> http://localhost:3001
    export ANTHROPIC_API_KEY=...
    uv run --with cortadel-langgraph --with "langchain[anthropic]" python examples/03_multi_user_graph.py
"""
from __future__ import annotations

import os
from typing import Annotated

from langchain.chat_models import init_chat_model
from langchain_core.messages import AnyMessage, HumanMessage
from langgraph.checkpoint.memory import InMemorySaver
from langgraph.graph import END, START, StateGraph
from langgraph.graph.message import add_messages
from typing_extensions import TypedDict

from cortadel_langgraph import CortadelMemory, CortadelStore, namespace_user_id

BASE_URL = os.environ.get("CORTADEL_URL", "http://localhost:3001")
API_KEY = os.environ.get("CORTADEL_API_KEY")

# One store for every user. The callable picks the Cortadel user id out of the namespace.
store = CortadelStore(BASE_URL, namespace_user_id(-1), api_key=API_KEY)

# `{user_id}` is filled from config["configurable"]["user_id"] on every call.
memory = CortadelMemory(store, namespace=("memories", "{user_id}"))

model = init_chat_model("anthropic:claude-sonnet-4-5")


class State(TypedDict):
    messages: Annotated[list[AnyMessage], add_messages]
    # `recall_node` writes here by default. No reducer, so each turn overwrites it and the
    # injected memory block never accumulates in state — the same schema create_react_agent
    # builds for itself when it has a pre_model_hook.
    llm_input_messages: list[AnyMessage]


def call_model(state: State) -> dict:
    # Use what recall handed us when it ran; fall back to the raw transcript otherwise.
    messages = state.get("llm_input_messages") or state["messages"]
    return {"messages": [model.invoke(messages)]}


builder = StateGraph(State)
builder.add_node("recall", memory.recall_node)  # before the model: inject relevant memories
builder.add_node("model", call_model)
builder.add_node("remember", memory.remember_node)  # after the turn: persist it to Cortadel
builder.add_edge(START, "recall")
builder.add_edge("recall", "model")
builder.add_edge("model", "remember")
builder.add_edge("remember", END)

graph = builder.compile(checkpointer=InMemorySaver())


def say(text: str, *, user_id: str, thread_id: str) -> None:
    print(f"\n[{user_id}] you: {text}")
    result = graph.invoke(
        {"messages": [HumanMessage(text)]},
        # thread_id scopes the short-term transcript; user_id scopes the long-term memory.
        # They are independent: a user has many threads, and memory crosses all of them.
        config={"configurable": {"thread_id": thread_id, "user_id": user_id}},
    )
    print(f"[{user_id}] agent: {result['messages'][-1].content}")


def main() -> None:
    alice = "e2e-langgraph-alice"
    bob = "e2e-langgraph-bob"

    if store.health(("memories", alice)) is None:
        raise SystemExit(f"No Cortadel server at {BASE_URL}. Start one with `docker compose up`.")

    say("I'm a vegetarian and I bike to work.", user_id=alice, thread_id="a1")
    say("I eat steak twice a week and I drive.", user_id=bob, thread_id="b1")

    print("\n(waiting for Cortadel to extract facts...)")
    import time

    time.sleep(5)

    # New threads. Each user's memories are namespaced to them, so neither answer can leak the
    # other's facts.
    say("Suggest somewhere for me to have lunch.", user_id=alice, thread_id="a2")
    say("Suggest somewhere for me to have lunch.", user_id=bob, thread_id="b2")


if __name__ == "__main__":
    try:
        main()
    finally:
        store.close()

File: integrations/langgraph/examples/03-multi-user-graph.ts
/**
 * One graph, many users — and no LLM at all, so this runs against nothing but a Cortadel server.
 *
 * A Cortadel client is bound to a single user id at construction, so per-user scoping means one
 * client per user. `namespaceUserId(-1)` is how one `CortadelStore` serves a multi-tenant graph:
 * it reads the user id out of the namespace, and the store keeps one client per resolved id.
 *
 * `thread_id` scopes the short-term transcript; `userId` scopes long-term memory. They are
 * independent — a user has many threads, and their memory crosses all of them.
 *
 * Run it with a Cortadel server on :3001 (`docker compose up` from the repo root):
 *
 *   pnpm dlx tsx examples/03-multi-user-graph.ts
 */
import process from "node:process";
import { AIMessage, HumanMessage, SystemMessage, type BaseMessage } from "@langchain/core/messages";
import {
  Annotation,
  END,
  MemorySaver,
  START,
  StateGraph,
  messagesStateReducer,
  type Messages,
} from "@langchain/langgraph";

import { CortadelMemory, CortadelStore, namespaceUserId } from "@cortadel/langgraph";

const BASE_URL = process.env.CORTADEL_BASE_URL ?? "http://localhost:3001";

// The two channels `createReactAgent` uses, declared by hand because this graph is hand-built.
// `llmInputMessages` is overwritten on every hook run rather than appended to, which is what keeps
// the recalled memory block out of the durable transcript.
const AgentState = Annotation.Root({
  messages: Annotation<BaseMessage[], Messages>({
    reducer: messagesStateReducer,
    default: () => [],
  }),
  llmInputMessages: Annotation<BaseMessage[], Messages>({
    reducer: (_left: BaseMessage[], right: Messages) => messagesStateReducer([], right),
    default: () => [],
  }),
});

/** Stands in for a model call, and reports whatever memory it was handed. */
function respond(state: typeof AgentState.State) {
  const seen = state.llmInputMessages.length > 0 ? state.llmInputMessages : state.messages;
  const block = seen.find((message) => message instanceof SystemMessage);
  return {
    messages: [new AIMessage(block ? `I recall: ${String(block.content)}` : "Nothing on file yet.")],
  };
}

async function main(): Promise<void> {
  const store = new CortadelStore({
    baseUrl: BASE_URL,
    // The user id comes out of the namespace's trailing label, so one store serves everybody.
    userId: namespaceUserId(-1),
    apiKey: process.env.CORTADEL_API_KEY,
  });

  // `{userId}` is filled from `config.configurable` on every call.
  const memory = new CortadelMemory({ store, namespace: ["memories", "{userId}"] });

  const graph = new StateGraph(AgentState)
    .addNode("recall", memory.recallNode)
    .addNode("respond", respond)
    .addNode("remember", memory.rememberNode)
    .addEdge(START, "recall")
    .addEdge("recall", "respond")
    .addEdge("respond", "remember")
    .addEdge("remember", END)
    .compile({ store, checkpointer: new MemorySaver() });

  for (const userId of ["e2e-langgraph-alice", "e2e-langgraph-bob"]) {
    // Each user gets their own thread and their own Cortadel namespace.
    await graph.invoke(
      { messages: [new HumanMessage(`My name is ${userId} and I work on the platform team.`)] },
      { configurable: { thread_id: `${userId}-t1`, userId } },
    );

    // A fresh thread for the same user: the transcript is empty, the memory is not.
    const result = await graph.invoke(
      { messages: [new HumanMessage("What do you know about me?")] },
      { configurable: { thread_id: `${userId}-t2`, userId } },
    );
    console.log(`${userId}: ${String(result.messages[result.messages.length - 1]?.content)}`);
  }
}

await main();
